### ISSUE TITLE
Introduce general naming interface

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#2693]
# This is a Feature Proposal

## Description

@dougmoscrop brought up an interesting idea to generalize the way naming (e.g. for resource logical ids) currently works.

See https://github.com/serverless/serverless/pull/2679#issuecomment-259707727 and https://github.com/serverless/serverless/pull/2679#issuecomment-259708354

Brought up in #2679

**Comment by erikerikson** [comment]
Importing @dougmoscrop 's comments:
> Would exposing a more general interface for naming make multi-provider support easier in the future?
> I'm thinking like
> this.provider.naming.sequential('Thing') => Thing1 (Then Thing2, etc.)
> this.provider.naming.resource('Thing{foo}') => ThingFooVar
> i.e. hide the 'rules' for how things get named behind some more common concepts.

And:
> You might also use a builder style pattern. For example, inside apiGateway you might do:
> this.naming = this.provider.naming.prefix('ApiGateway');
> and then inside each apiGateway compiler, you can just do (e.g.) this.naming.sequential('ApiKey')
> which would produce ApiGatwayApiKey1
> It feels like a similar domain to structured logging with the like 'child namers'.

**Comment by erikerikson** [comment]
Personally, I'd be wary of abstracting away providers.  If we could provide global naming conventions and then override those on a specific provider by provider basis, my worries on this issue with disappear.

The basis for my concern is that I know a lot of our naming standards requirement evolutions have been the result of vendor and vendor service specific nuances and we'll need to be empowered to deal with those by any naming specification capability.

**Comment by erikerikson** [comment]
our requirements in the past have relied heavily upon variable usage in order to compact and generalize our conventions.  it will be very important for us to be able to use variables in any solution that will be viable for our use.

**Comment by dougmoscrop** [comment]
When I wrote those comments I was thinking about it from the perspective of
someone maybe writing a plugin that would add resources and need to name
them. I'm wondering now if that's ever going to be a thing, and even if it
is, such a plugin would probably be provider specific anyway..

**Comment by erikerikson** [comment]
Interesting perspective.  My need is to be able to fully control the deployed name prefixes of assets deployed by the framework (i.e. we don't care about logical IDs).  Maybe there are a set of use cases that can be identified and addressed.  It would be nice if plugin's would magically be able to incorporate naming configuration but I don't have any burning ideas about how that could be implemented.  Will ponder.

**Comment by erikerikson** [comment]
For the record, I am hoping to begin implementing something to cover our needs very soon.

**Comment by pmuens** [comment]
@erikerikson Awesome!

@eahefnawy what do you think about this feature?

**Comment by erikerikson** [comment]
TL;DR: Status Change: Non-Blocker, Low Priority

It turns out that an internal shift in cloud strategy has provided us a simpler path forward that allows us to adopt v1.0 in our new context and support the old context with v0.5.6.  I am officially withdrawing my intention to implement something here.  Apologies for the shift.

To clarify, we no longer have this feature as a blocker to our adoption.  That said, it will always remain nice to be able to control the information by which we will understand our system when searching for and reviewing its components from the console.  It's never been clear to me why it was important for the framework to have an opinion on the matter outside of a set of invariant conditions that avoid conflicts and other problem states.

**Comment by pmuens** [comment]
A general naming interface was introduced with the `naming.js` file: https://github.com/serverless/serverless/blob/master/lib/plugins/aws/lib/naming.js
